File: feature/collect/src/test/java/com/shang/jetpackmoviekmp/feature/collect/ui/FakeMovieRepository.kt
package com.shang.jetpackmoviekmp.feature.collect.ui

import androidx.paging.PagingData
import com.shang.jetpackmoviekmp.common.AppResult
import com.shang.jetpackmoviekmp.data.repository.MovieRepository
import com.shang.jetpackmoviekmp.model.ConfigurationBean
import com.shang.jetpackmoviekmp.model.MovieCardResult
import com.shang.jetpackmoviekmp.model.MovieCastAndCrewBean
import com.shang.jetpackmoviekmp.model.MovieDetailBean
import com.shang.jetpackmoviekmp.model.MovieGenreBean
import com.shang.jetpackmoviekmp.model.MovieRecommendBean
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.flowOf
import java.util.concurrent.CountDownLatch

internal class FakeMovieRepository : MovieRepository {

    val collectedMovies = MutableStateFlow<List<MovieCardResult>>(emptyList())
    var deleteMovieCollectCallCount = 0
        private set
    var deletedMovie: MovieCardResult? = null
        private set
    val deleteMovieCollectLatch = CountDownLatch(1)

    override fun getConfiguration(): Flow<Result<ConfigurationBean>> = flowOf(Result.success(ConfigurationBean()))

    override fun getMovieGenres(): Flow<AppResult<MovieGenreBean>> = flowOf(AppResult.Success(MovieGenreBean()))

    override fun getMovieListPager(withGenres: String): Flow<PagingData<MovieCardResult>> =
        flowOf(PagingData.empty())

    override fun getMovieSearchPager(query: String): Flow<PagingData<MovieCardResult>> =
        flowOf(PagingData.empty())

    override fun getMovieDetail(id: Int): Flow<Result<MovieDetailBean>> = flowOf(Result.success(MovieDetailBean()))

    override fun getMovieRecommendations(id: Int): Flow<Result<MovieRecommendBean>> =
        flowOf(Result.success(MovieRecommendBean()))

    override fun getMovieActor(id: Int): Flow<AppResult<MovieCastAndCrewBean>> =
        flowOf(AppResult.Success(MovieCastAndCrewBean()))

    override suspend fun insertMovieCollect(movieResult: MovieCardResult) = Unit

    override suspend fun deleteMovieCollect(movieResult: MovieCardResult) {
        deleteMovieCollectCallCount++
        deletedMovie = movieResult
        deleteMovieCollectLatch.countDown()
    }

    override fun getCollectedMovieIds(): Flow<List<Int>> = flowOf(emptyList())

    override fun getAllMovieCollect(): Flow<List<MovieCardResult>> = collectedMovies

    override fun getMovieCollectEntityById(id: Int): Flow<MovieCardResult?> = flowOf(null)

    override suspend fun insertMovieHistory(movieResult: MovieCardResult) = Unit

    override suspend fun deleteMovieHistory(movieResult: MovieCardResult) = Unit

    override fun getAllMovieHistory(): Flow<List<MovieCardResult>> = flowOf(emptyList())

    override suspend fun deleteAllMovieHistory(): Boolean = false
}

File: iosApp/iosAppTests/Common/FakeMovieRepository.swift
import Shared

/// `MovieRepository` 是完整的 Kotlin interface；這個 Fake 只提供編譯所需的最小實作，
/// 讓測試能建構持有 `MovieRepository` 的 ViewModel，測試本身不驗證這裡的行為。
///
/// 只需實作 `async throws` 版本：Swift 會自動為其產生對應的 completion-handler
/// ObjC 進入點，滿足協定要求的兩個 selector；若同時手動實作兩者會導致 ObjC selector 衝突。
final class FakeMovieRepository: MovieRepository {
    func getConfiguration() -> SkieSwiftFlow<Any> {
        fatalError("not used in these tests")
    }

    func getMovieGenres() -> SkieSwiftFlow<any AppResult> {
        fatalError("not used in these tests")
    }

    func getMovieListPager(withGenres: String) -> SkieSwiftFlow<Paging_commonPagingData<MovieCardResult>> {
        fatalError("not used in these tests")
    }

    func getMovieSearchPager(query: String) -> SkieSwiftFlow<Paging_commonPagingData<MovieCardResult>> {
        fatalError("not used in these tests")
    }

    func getMovieDetail(id: Int32) -> SkieSwiftFlow<Any> {
        fatalError("not used in these tests")
    }

    func getMovieRecommendations(id: Int32) -> SkieSwiftFlow<Any> {
        fatalError("not used in these tests")
    }

    func getMovieActor(id: Int32) -> SkieSwiftFlow<any AppResult> {
        fatalError("not used in these tests")
    }

    func __insertMovieCollect(movieResult: MovieCardResult) async throws {
        fatalError("not used in these tests")
    }

    func __deleteMovieCollect(movieResult: MovieCardResult) async throws {
        fatalError("not used in these tests")
    }

    func getCollectedMovieIds() -> SkieSwiftFlow<[KotlinInt]> {
        fatalError("not used in these tests")
    }

    func getAllMovieCollect() -> SkieSwiftFlow<[MovieCardResult]> {
        fatalError("not used in these tests")
    }

    func getMovieCollectEntityById(id: Int32) -> SkieSwiftOptionalFlow<MovieCardResult> {
        fatalError("not used in these tests")
    }

    func getAllMovieHistory() -> SkieSwiftFlow<[MovieCardResult]> {
        fatalError("not used in these tests")
    }

    func __insertMovieHistory(movieResult: MovieCardResult) async throws {
        fatalError("not used in these tests")
    }

    func __deleteMovieHistory(movieResult: MovieCardResult) async throws {
        fatalError("not used in these tests")
    }

    func __deleteAllMovieHistory() async throws -> KotlinBoolean {
        fatalError("not used in these tests")
    }
}
